File: Chapter 09/C#/section_9_3_4_serverless_functions/Handlers/AzureFunctionHandler.cs
using System;
using System.IO;
using Chapter09.ServerlessFunctions.Infrastructure.Azure;

namespace Chapter09.ServerlessFunctions.Handlers;

public class AzureFunctionHandler
{
    /// <summary>
    /// CLOUD 2: AZURE FUNCTIONS (The Declarative App)
    /// 
    /// THE ARCHITECTURAL LESSON: 
    /// Azure uses "Bindings" to abstract away the network plumbing, but it 
    /// "Owns" your method signature in exchange for that convenience.
    /// 
    /// TEACHING NOTE:
    /// Look at the 'Stream myBlob' parameter. Unlike the AWS example, we don't 
    /// receive a JSON event; we receive the actual file content! Azure's host 
    /// process performed the download for us before our code even started. 
    /// While this is "cleaner," notice the trade-off: this method signature 
    /// is now proprietary to the Azure Functions runtime. The infrastructure 
    /// has "Leaked" into our method arguments, making it difficult to run 
    /// this code anywhere else without a wrapper.
    /// </summary>
    
    // In real production code, you would see: 
    // [FunctionName("ResizeImage")]
    // public static void Run([BlobTrigger("uploads/{name}")] Stream myBlob...)
    public string Handle(Stream myBlob, string name, MockAzureLogger log)
    {
        // 1. THE CLOUD CONTRACT & DECLARATIVE FETCH: Combined by the platform
        // The data is already present in a Stream object handed to us.
        long fileSize = myBlob.Length;

        log.LogInformation($"      [Azure Function] Stream injected via bindings: {name}");
        log.LogInformation($"      [Azure Function] Azure handled the network fetch automatically.");
        
        // 2. THE LOGIC:
        log.LogInformation($"      [Azure Function] Processing image resize...");

        // 3. THE RESPONSE: Azure allows simple return types that map back to the cloud
        return $"Azure processed {name}";
    }
}

File: Chapter 09/JS/section_9_3_4_serverless_functions/handlers/azureFunctionHandler.js
import { map } from 'rxjs/operators';

/**
 * CLOUD 2: AZURE FUNCTIONS (The Declarative App - Reactive)
 * * THE ARCHITECTURAL LESSON: 
 * Azure uses "Bindings" to abstract network plumbing. Your code doesn't 
 * "fetch" data; it "receives" data. However, the platform "Owns" your 
 * function's signature.
 * * @param {MockAzureContext} context - The proprietary Azure context.
 * @returns {Observable} A stream emitting the result of the mutation.
 */
export const execute$ = (context) => {
    return context.asObservable$().pipe(
        /**
         * TEACHING NOTE:
         * Notice the difference from AWS. We don't have to map a 'key' 
         * to a manual SDK call. The 'blobData' is already here. 
         * But look at the return: we are mutating 'context.res'. 
         * This "Platform Contamination" makes it nearly impossible to 
         * test this reactive pipe without bringing the whole Azure 
         * Mock object with it. The infrastructure has leaked into the 
         * very heart of the stream.
         */
        map(ctx => {
            const fileName = ctx.bindingData.name;
            const fileSize = ctx.blobData ? ctx.blobData.length : 0; // Safe access

            ctx.log(`      [Azure Function] Stream observed: ${fileName} (${fileSize} bytes)`);
            ctx.log(`      [Azure Function] Declarative logic: Processing resize...`);

            ctx.res = {
                status: 200,
                body: `Azure reactive processed ${fileName}`
            };

            return ctx.res;
        })
    );
};